packages/system/src/system.ts

```diff
@@ -38,20 +46,38 @@ interface GetStyleObject {
  * behaviors. Right now, the `sx` prop has the highest priority so the resolved
  * fontSize will be `40px`
  */
-export const toCSSObject: GetStyleObject = ({ baseStyle }) => (props) => {
-  const { theme, css: cssProp, __css, sx, ...rest } = props
-  const styleProps = objectFilter(rest, (_, prop) => isStyleProp(prop))
-  const finalBaseStyle = runIfFn(baseStyle, props)
-  const finalStyles = Object.assign(
-    {},
-    __css,
-    finalBaseStyle,
-    filterUndefined(styleProps),
-    sx,
-  )
-  const computedCSS = css(finalStyles)(props.theme)
-  return cssProp ? [computedCSS, cssProp] : computedCSS
-}
+export const toCSSObject: GetStyleObject =
+  ({ baseStyle }) =>
+  (props) => {
+    const { theme, css: cssProp, __css, sx, ...rest } = props
+
+    const [otherProps, imdStyleProps]: [
+      otherProps: Config,
+      styleDomProps: Config,
+    ] = split(rest, othersPropNames)
+    const styleProps: Config = filterUndefined(
+      pick(imdStyleProps, stylePropNames),
+    )
+    const stylePrecedence: any[] = []
+
+    // Styles are applied by precedence (lowest to highest)
+    //   __css
+    //   baseStyle
+    //   other (apply, layerStyles, textStyles)
+    //   styleProps
+    //   sx
+    stylePrecedence.push(__css)
+    stylePrecedence.push(runIfFn(baseStyle, props))
+    stylePrecedence.push(otherProps)
+    stylePrecedence.push(styleProps)
+    stylePrecedence.push(sx)
+
+    const computedCSS = stylePrecedence.reduce((accStyle: any, style: any) => {
+      return mergeWith(accStyle, css(style)(theme))
+    }, {})
```

Comment:
The `mergeWith` function causes to increase the runtime complexity to increase from O(n) to O(n^2). Sadly this is in the `styled-system` algorithm hot path, which will be called by every chakra component on every render. 

A very simple loop test gave us the following average time in ms for a run with 10000 iterations:

PR solution: 0.045435ms
Current main branch: 0.02831ms

The large chunk of the performance regression is caused by running `mergeWith` and `css` function in a loop, compared to the current approach that uses `Object.assign` and runs the `css` function once.